### PR TITLE
Bound check elimination using array.size

### PR DESCRIPTION
**Pull request by soywiz**
Proposal. Might have bugs. Let's wait for tests to run :)

Tested with this snippet:
```
fun loop(out: IntArray) {
        for (n in 0 until out.size) {
                out[n] = 10
        }
}

fun main(args: Array<String>) {
        val out = IntArray(1024)
        loop(out)
        println(out[10])
}
```

This should fix: https://github.com/JetBrains/kotlin-native/issues/1791

The idea is that K/N was generating for a:

```
for (n in 0 until array.size) X
```

a loop with (simplified):

```
val last = array.size - 1
var n = 0
do {
   X
} (n++ != last)
```

and LLVM doesn't like that.

Now I'm generating:

```
val last = array.size 
var n = 0
do {
   X
} (n++ < last)
```

And LLVM likes it and can detect that last matches the check in the bounds checking.
It still checks if the array size <= 0 before the loop starts. It should be fixed if the code is changed to:

```
val last = array.size 
var n = 0
while (n < last) {
   X
   n++
}
```

Also calling the `Kotlin_*_getArrayLength` directly + const was required for LLVM to like this.

### CODE REVIEW

File: runtime/src/main/kotlin/kotlin/Arrays.kt [hunk 1]
```diff
@@ -146,12 +148,12 @@ public final class ShortArray {
...
     @InlineConstructor
     public constructor(size: Int, init: (Int) -> Short): this(size) {
-        for (i in 0..size - 1) {
+        for (i in 0 until this.size) {
             this[i] = init(i)
         }
     }
 
-    public val size: Int
+    inline public val size: Int
```
Comment (olonho) [inline runtime/src/main/kotlin/kotlin/Arrays.kt]: Better be handled by codegen, I guess, and actually may make optimizer understand what's going on easier.

**Comment by sbogolepov** [comment]
Sadly, it is not working in ugly corner-cases like this one:
```kotlin
fun main(args: Array<String>) {
    for (i in Int.MAX_VALUE..Int.MAX_VALUE) {
        println(i)
    }
}
```
Still, the proposal is great and we will investigate ways to improve it. Suggestions are welcome :)

File: runtime/src/main/kotlin/kotlin/Arrays.kt (hunk 1 above)
Comment (soywiz) [reply runtime/src/main/kotlin/kotlin/Arrays.kt]: Before doing this PR I tried to some intrinsics but I had some strange errors about a symbol not found, so maybe I did it wrong.

**Comment by soywiz** [comment]
@sbogolepov Yeah :( , I have just checked and understood the problem. Though for until + step 1, this problem shouldn't happen. I have tried to do a quick change but would require some adjustments.
Maybe it could be optimized for some cases only where you know there is not going to be integer overflow.

So I'm going to close this for now.

Right now I'm waiting for this: https://github.com/JetBrains/kotlin-native/issues/1780 to be fixed and I will be able to run a performance-critical app, after that I will be able to check again and fine-tune, and I'll revisit this with more information :) and will make another PR if I manage to cover all the cases and improves the performance and code size. Next time I will also try edge cases at the bounds with several steps beforehand. :+1:

**Comment by sbogolepov** [comment]
BTW, running `:backend.native:tests:run_external` is a good way to detect mistakes :)
